src/torchmetrics/functional/classification/group_fairness.py

```diff
@@ -0,0 +1,366 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Dict, Optional, Tuple
+
+import torch
+from typing_extensions import Literal
+
+from torchmetrics.functional.classification.stat_scores import (
+    _binary_stat_scores_arg_validation,
+    _binary_stat_scores_format,
+    _binary_stat_scores_tensor_validation,
+    _binary_stat_scores_update,
+)
+from torchmetrics.utilities import rank_zero_warn
+from torchmetrics.utilities.compute import _safe_divide
+
+
+def _groups_validation(groups: torch.Tensor, num_groups: int) -> None:
+    """Validate groups tensor.
+
+    - The largest number in the tensor should not be larger than the number of groups. The group identifiers should
+    be ``0, 1, ..., (num_groups - 1)``.
+    - The group tensor should be dtype long.
+    """
+    if torch.max(groups) > num_groups:
+        raise ValueError(
+            f"The largest number in the groups tensor is {torch.max(groups)}, which is larger than the specified",
+            f"number of groups {num_groups}. The group identifiers should be ``0, 1, ..., (num_groups - 1)``.",
+        )
+    if groups.dtype != torch.long:
+        raise ValueError(f"Excpected dtype of argument groups to be long, not {groups.dtype}.")
+
+
+def _groups_format(groups: torch.Tensor) -> torch.Tensor:
+    """Reshape groups to correspond to preds and target."""
+    groups = groups.reshape(groups.shape[0], -1)
+
+    return groups
+
+
+def _binary_groups_stat_scores(
+    preds: torch.Tensor,
+    target: torch.Tensor,
+    groups: torch.Tensor,
+    num_groups: int,
+    threshold: float = 0.5,
+    ignore_index: Optional[int] = None,
+    validate_args: bool = True,
+) -> Dict[str, Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]]:
+    """Computes the number of true positives, false positives, true negatives, and false negatives for binary
+    classification by group.
+
+    Related to `Type I and Type II errors`_.
+    """
+    if validate_args:
+        _binary_stat_scores_arg_validation(threshold, "global", ignore_index)
+        _binary_stat_scores_tensor_validation(preds, target, "global", ignore_index)
+        _groups_validation(groups, num_groups)
+
+    preds, target = _binary_stat_scores_format(preds, target, threshold, ignore_index)
+    groups = _groups_format(groups)
+
+    group_preds = [preds[groups == i].unsqueeze(1) for i in range(num_groups)]
+    group_target = [target[groups == i].unsqueeze(1) for i in range(num_groups)]
+
```

Comment:
This looks inefficient because you are indexing many times the same vector. I suggest taking a look at the retrieval metrics, in which a similar problem is solved with a single indexing.
Take a look here: https://github.com/Lightning-AI/metrics/blob/d2ef383c16f52f02ff38e7c7b8146010e694036b/src/torchmetrics/retrieval/base.py#L119

Reply:
Thanks for the example! I believe I've used a similar technique now to use single indexing.